File: csharp/behavioral/observer.cs
namespace Gof.ModernCSharp.Behavioral;

// Observer notifies subscribers whenever the subject publishes a change.
// Modern C# note: events, IObservable<T>, channels, and message buses are common.
// Be explicit about unsubscription in long-lived systems.
public static class Observer
{
    sealed class NewsFeed
    {
        public event Action<string>? Published;
        public void Publish(string headline) => Published?.Invoke(headline);
    }

    public static string Run()
    {
        var messages = new List<string>();
        var feed = new NewsFeed();
        feed.Published += headline => messages.Add($"email: {headline}");
        feed.Published += headline => messages.Add($"sms: {headline}");
        feed.Publish("Pattern examples released");
        return string.Join(", ", messages);
    }
}

File: py/behavioral/observer.py
# Observer notifies subscribed callbacks whenever the subject publishes a change.
# Modern Python note: callbacks, signals, async queues, and pub/sub libraries are the
# usual forms. Be explicit about ownership and unsubscription in long-lived systems.
class NewsFeed:
    def __init__(self):
        self.subscribers = []

    def subscribe(self, callback):
        self.subscribers.append(callback)

    def publish(self, headline):
        return [callback(headline) for callback in self.subscribers]


def email_subscriber(headline):
    return f"email: {headline}"


def sms_subscriber(headline):
    return f"sms: {headline}"


if __name__ == "__main__":
    feed = NewsFeed()
    feed.subscribe(email_subscriber)
    feed.subscribe(sms_subscriber)
    print(feed.publish("Pattern examples released"))
